File: app/src/main/java/de/mein/android/drive/bash/SAFBashTools.java
package de.mein.android.drive.bash;

import android.content.Context;

import org.jdeferred.Promise;

import java.io.File;
import java.io.IOException;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;

import de.mein.android.file.AndroidFileConfiguration;
import de.mein.auth.file.AFile;
import de.mein.drive.bash.BashToolsException;
import de.mein.drive.bash.BashToolsImpl;
import de.mein.drive.bash.FsBashDetails;

public class SAFBashTools implements BashToolsImpl {
    private final Context context;
    private AndroidFileConfiguration configuration;
    private BashToolsAndroid bashToolsAndroid;

    public SAFBashTools(Context context) {
        this.context = context;
        AFile.Configuration configuration = AFile.getConfiguration();
        if (configuration == null) {
            System.err.println("SAFBashTools.SAFBashTools: " + AFile.class.getSimpleName() + " was not configured. Call " + AFile.class.getSimpleName() + ".configure() before calling this.");
        } else if (!(configuration instanceof AndroidFileConfiguration)) {
            System.err.println("SAFBashTools.SAFBashTools: " + AFile.class.getSimpleName() + " was improperly configured. Configuration type is " + configuration.getClass().getSimpleName() + " but " + AndroidFileConfiguration.class.getSimpleName() + " was expected. Call " + AFile.class.getSimpleName() + ".configure() before calling this.");
        } else {
            this.configuration = (AndroidFileConfiguration) AFile.getConfiguration();
            bashToolsAndroid = new BashToolsAndroid(context);
        }
    }


    @Override
    public void setBinPath(String binPath) {
        //nothing to do here
    }

    @Override
    public Set<Long> getINodesOfDirectory(AFile file) throws IOException {
//        DFile dFile = (DFile) file;
//        if (dFile.isRawFile())
//            return bashToolsAndroid.getINodesOfDirectory(dFile);
//
//        configuration.getContext().getContentResolver().query(dFile.getUri(), DocumentsContract.Document.COLUMN_DISPLAY_NAME,)
        return null;
    }

    @Override
    public FsBashDetails getFsBashDetails(AFile file) throws IOException, InterruptedException {
        return null;
    }

    @Override
    public void rmRf(AFile directory) throws IOException {

    }

    @Override
    public List<AFile> stuffModifiedAfter(AFile referenceFile, AFile directory, AFile pruneDir) throws IOException, BashToolsException {
        return null;
    }

    @Override
    public Iterator<AFile<?>> find(AFile directory, AFile pruneDir) throws IOException {
        //testing needed which one is faster.
        return bashToolsAndroid.find(directory, pruneDir);
//        return new DFileRecursiveIterator(context, directory, pruneDir);
    }

    @Override
    public Iterator<AFile> stuffModifiedAfter(AFile originalFile, AFile pruneDir, long timeStamp) throws IOException, InterruptedException {
        return null;
    }

    @Override
    public void mkdir(AFile dir) throws IOException {

    }

    @Override
    public boolean mv(File source, File target) throws IOException {
        bashToolsAndroid.mv(source, target);
        return false;
    }

    @Override
    public boolean isSymLink(AFile f) {
        return false;
    }

    @Override
    public Map<String, FsBashDetails> getContentFsBashDetails(AFile file) {
        return bashToolsAndroid.getContentFsBashDetails(file);
    }

    @Override
    public void lnS(AFile file, String target) {
        // symlinks do not work on Android
    }
}

File: android/app/src/main/java/de/mel/android/filesync/bash/SAFBashTools.kt
package de.mel.android.filesync.bash

import android.content.ContentResolver
import android.database.Cursor
import android.net.Uri
import android.os.Build
import android.provider.DocumentsContract
import android.provider.OpenableColumns
import androidx.annotation.RequiresApi
import de.mel.Lok
import de.mel.android.file.AndroidFile
import de.mel.android.file.AndroidFileConfiguration
import de.mel.android.file.SAFFile
import de.mel.android.file.SAFFileConfiguration
import de.mel.auth.file.AbstractFile
import de.mel.filesync.bash.AutoKlausIterator
import de.mel.filesync.bash.BashTools
import de.mel.filesync.bash.BashToolsException
import de.mel.filesync.bash.FsBashDetails
import java.io.File
import java.util.*

@RequiresApi(Build.VERSION_CODES.Q)
class SAFBashTools : BashTools<SAFFile>() {
    override fun getFsBashDetails(file: SAFFile): FsBashDetails? {
        val docFile = file.getDocFile()
        val uri = docFile!!.uri
        val contentResolver: ContentResolver = (AbstractFile.configuration as SAFFileConfiguration).context.contentResolver
        contentResolver.query(uri, arrayOf(DocumentsContract.Document.COLUMN_DOCUMENT_ID, DocumentsContract.Document.COLUMN_LAST_MODIFIED, DocumentsContract.Document.COLUMN_DISPLAY_NAME), null, null, null)
                ?.use {
                    it.moveToFirst()
                    val modified = it.getLong(it.getColumnIndex(DocumentsContract.Document.COLUMN_LAST_MODIFIED))
                    val inode = it.getLong(it.getColumnIndex(DocumentsContract.Document.COLUMN_DOCUMENT_ID))
                    val name = it.getString(it.getColumnIndex(DocumentsContract.Document.COLUMN_DISPLAY_NAME))
                    val fsBashDetails = FsBashDetails(modified, modified, inode, false, null, name)
                    return fsBashDetails
                }
        return null
    }


    override fun lnS(file: SAFFile, target: String) {
        throw BashToolsException("Android does not support symbolic links. This method must not be called!")
    }

    override fun getContentFsBashDetails(dir: SAFFile): MutableMap<String, FsBashDetails> {
        try {
            val thisDoc = dir.getDocFile()!!
            val uri: Uri = DocumentsContract.buildChildDocumentsUriUsingTree(thisDoc.uri, DocumentsContract.getDocumentId(thisDoc.uri))
            val contentResolver: ContentResolver = (AbstractFile.configuration as SAFFileConfiguration).context.contentResolver
            val content = mutableMapOf<String, FsBashDetails>()
            contentResolver.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME
                    , DocumentsContract.Document.COLUMN_LAST_MODIFIED
                    , DocumentsContract.Document.COLUMN_DOCUMENT_ID), null, null, null, null)?.use {
                it.moveToFirst()
                val nameIndex = it.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                val modIndex = it.getColumnIndex(DocumentsContract.Document.COLUMN_LAST_MODIFIED)
                val docIdIndex = it.getColumnIndex(DocumentsContract.Document.COLUMN_DOCUMENT_ID)
                while (!it.isAfterLast) {
                    val name = it.getString(nameIndex)
                    val modified = it.getLong(modIndex)
                    val docId = it.getLong(docIdIndex)
                    val details = FsBashDetails(modified, modified, docId, false, null, name)
                    content[name] = details
                    it.moveToNext()
                }
            }
            return content
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return mutableMapOf()
    }

    override fun rmRf(directory: SAFFile) {
    }

    override fun stuffModifiedAfter(referenceFile: SAFFile, directory: SAFFile, pruneDir: SAFFile): List<SAFFile> {
        throw BashToolsException("Method does not work on Android 10+ anymore.")
    }

    private class SAFFileHelper(val name: String, val mime: String) {
        fun isDir() = mime == DocumentsContract.Document.MIME_TYPE_DIR
    }

    override fun find(directory: SAFFile, pruneDir: SAFFile): AutoKlausIterator<SAFFile> {
        // todo saf the code currently ignores pruneDir until I found a way to recognize it via path without fucking with the documentfile cache
        // DocFileCache works best using DFS. Another thing SAF fucks up.
        val prunePath = pruneDir.absolutePath
        val config = AbstractFile.configuration as SAFFileConfiguration
        val cache = config.externalCache
        val docFile = directory.getDocFile()
        val uri = docFile!!.uri
        val contentResolver: ContentResolver = (AbstractFile.configuration as SAFFileConfiguration).context.contentResolver
        // Find does DFS, ContentResolver does not. What a nice day.

        contentResolver.query(directory.getChildrenUri(), arrayOf(DocumentsContract.Document.COLUMN_DISPLAY_NAME, DocumentsContract.Document.COLUMN_MIME_TYPE), null, null, null)
                ?.use { cursor ->
                    cursor.moveToFirst()
                    val nameIndex = cursor.getColumnIndex(DocumentsContract.Document.COLUMN_DISPLAY_NAME)
                    val mimeIndex = cursor.getColumnIndex(DocumentsContract.Document.COLUMN_MIME_TYPE)
                    val klaus = object : AutoKlausIterator<SAFFile> {
                        /**
                         * Will keep an Iterator of sorted folder content for each step down.
                         */
                        lateinit var cursor: Cursor
                        val fileStack = Stack<SAFFile>()
                        val sortedFilesStack = Stack<Iterator<SAFFile>>()
                        override fun hasNext(): Boolean = sortedFilesStack.size > 0 && sortedFilesStack.peek().hasNext()

                        init {
                            fileStack.push(directory)
                        }

                        override fun next(): SAFFile {
                            while (!sortedFilesStack.peek().hasNext())
                                sortedFilesStack.pop()
                            return sortedFilesStack.peek().next()
                        }

                        fun readCursor(cursor: Cursor): Unit {
                            this.cursor = cursor
                            val parentFile = fileStack.peek()
                            val pathParts = AndroidFile.createRelativeFilePathParts(parentFile.storagePath, File(parentFile.absolutePath))
                            val list = mutableListOf<SAFFile>()
                            cursor.use {
                                it.moveToFirst()

                                while (!it.isAfterLast) {
                                    val mime = cursor.getString(mimeIndex)
                                    val name = cursor.getString(nameIndex)

                                    val file = AbstractFile.instance(parentFile, name) as SAFFile
                                    // filter prune directory
                                    if (!file.absolutePath.startsWith(prunePath))
                                        list.add(file)
                                    it.moveToNext()
                                }
                            }
                            // todo saf check if the contentresolver can sort the results
                            list.sortWith(Comparator { a, b ->
                                when {
                                    a.isDirectory == b.isDirectory -> a.getName().compareTo(b.getName())
                                    a.isDirectory -> -1
                                    else -> 1
                                }
                            })
//                            config.internalCache!!.findDoc(pathParts)
                            sortedFilesStack.push(list.iterator())
                        }

                        override fun close() {
                            Lok.debug("closed")
                        }
                    }
                    klaus.readCursor(cursor)
                    return klaus
                }
        return AutoKlausIterator.EmptyAutoKlausIterator()
    }

    override fun stuffModifiedAfter(directory: SAFFile, pruneDir: SAFFile, timeStamp: Long): AutoKlausIterator<SAFFile> {
        return AutoKlausIterator.EmptyAutoKlausIterator()
    }
}